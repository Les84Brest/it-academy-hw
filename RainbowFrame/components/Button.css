.btn{
  font: 17px/27px Arial, sans-serif;
  background-color: #f57413;
  cursor: pointer;
  background: -webkit-linear-gradient(top,#ed8414, #f57413);
  background: -moz-linear-gradient(top,#ed8414, #f57413);
  background: -ms-linear-gradient(top,#ed8414, #f57413);
  background: -o-linear-gradient(top,#ed8414, #f57413);
  padding: 2px 7px;
  margin: 10px 0 5px;
  position: relative;
  box-shadow: -webkit-box-shadow -1px 1px 2px rgba(0,0,0,0.28), inset 0 1px 0 0 #ffa854;
  box-shadow: -moz-box-shadow -1px 1px 2px rgba(0,0,0,0.28),inset 0 1px 0 0 #ffa854;
  box-shadow: -1px 1px 1px rgba(0,0,0,0.28), 	inset 0 1px 0 0 #ffa854;
  border: 1px solid #d86c1b;
  display: block;
  text-transform: uppercase;
  color: #fff;
  min-width: 107px;
  text-shadow: 1px 1px 1px rgba(0,0,0, 0.3);
  line-height: 40px;
  text-decoration: none;
  border-radius: 5px;
  vertical-align: middle;
  text-align: center;
  }

.btn:hover{
  background: #ff911b;
}
.btn:active{
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.06);
  transform: scale(.9,.9);
  -moz-transform: scale(.9,.9);
  -webkit-transform: scale(.9,.9);
  -o-transform: scale(.9,.9);
  -ms-transform: scale(.9,.9);
  background: #ff911b;
}
.btn:disabled{
  opacity: 0.3;
}