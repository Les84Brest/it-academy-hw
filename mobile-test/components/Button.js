
import React, {Children, Component} from "react";
import PropTypes from 'prop-types';
import ClassNames from 'classnames';

// import './Button.css';



const Button = ({children, onClick, className, disabled, active, ...attrs}) => {
  
  const classes = ClassNames(
    'btn',
    className,
    {active},
  );

  const onClickAction = e => {
    if(disabled){
      e.preventDefault();
    }else{
      return onClick(e);
    }
  }
    const Tag = attrs.href ? 'a' : 'button';
  return(
    <Tag
    {...attrs}
    className={classes}
    onClick={onClickAction}
    disabled={disabled}
    >{children}</Tag>
  );
}

Button.propTypes = {
  children: PropTypes.node,
  onClick: PropTypes.func,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  active: PropTypes.bool,
};

Button.defaultProps = {
  children: 'DefaultButton',
  onClick : () => {},
  className : '',
  disabled: false,
  active: false,
}
export default Button;
