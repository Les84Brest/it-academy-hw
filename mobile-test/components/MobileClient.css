 tr.MobileClient td {
   border-right: 1px solid #dfdfdf;
 }

 tr.MobileClient td:last-child {
   border-right: 0px;
 }

 tr.MobileClient td.active{
   background-color: darkgreen;
 }
 tr.MobileClient td.blocked{
   background-color: tomato;
 }
/* добавление нового клиента */
 .MobileClient-new{
   color: tomato;
 }