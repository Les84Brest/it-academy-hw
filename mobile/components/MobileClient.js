import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Button from './Button';

import mobileEvents from './mobileEvents';

import './MobileClient.css';



class MobileClient extends React.PureComponent {

  static WORKMODE_SHOW = 1;
  static WORKMODE_EDIT = 2;
  static WORKMODE_NEW = 3;

  constructor(props) {
    super(props);
    // управляем режимом компонента


    if (props.clientInfo == null) {
      this.state.workMode = MobileClient.WORKMODE_NEW;
      // build new client hash for state

      this.state.clientInfo = {
        lastName: '',
        firstName: '',
        secondName: '',
        balanse: '',
        id: Math.floor(Math.random() * 100000),
        status: '',
      };
    }

    //refs

    this.firstNameRef = null;
    this.lastNameRef = null;
    this.secondNameRef = null;
    this.balanseRef = null;
    
    //new client
    this.firstNameNewRef = null;
    this.lastNameNewRef = null;
    this.secondNameNewRef = null;
    this.balanseNewRef = null;
    

  }

  // ref callbacks


  setLastNameRef = (ref) => {
    this.lastNameRef = ref;
  }
  setSecondNameRef = (ref) => {
    this.secondNameRef = ref;
  }
  setFirstNameRef = (ref) => {
    this.firstNameRef = ref;
  }
  setBalanseRef = (ref) => {
    this.balanseRef = ref;
  }
// new client ref
  setLastNameNewRef = (ref) => {
    this.lastNameNewRef = ref;
  }
  setSecondNameNewRef = (ref) => {
    this.secondNameNewRef = ref;
  }
  setFirstNameNewRef = (ref) => {
    this.firstNameNewRef = ref;
    console.log(this.firstNameNewRef); 
  }
  setBalanseNewRef = (ref) => {
    console.log("ref balanse new");
    this.balanseNewRef = ref;
    console.log(this.balanseNewRef);
  }


  static propTypes = {

    clientInfo: PropTypes.shape({
      lastName: PropTypes.string,
      firstName: PropTypes.string,
      secondName: PropTypes.string,
      balanse: PropTypes.number,
      id: PropTypes.number,
      status: PropTypes.string,
    }
    )

  }

  state = {
    clientInfo: this.props.clientInfo,
    workMode: MobileClient.WORKMODE_SHOW,
  }


  componentDidUpdate = () =>{
    
    this.setState({clientInfo: this.props.clientInfo});
  }

  handlerSaveNewClient = () => {

    let client = { ...this.state.clientInfo };
   console.log(this.balanseNewRef.value); 

    client.lastName = this.lastNameNewRef.value;
    client.firstName = this.firstNameNewRef.value;
    client.secondName = this.secondNameNewRef.value;
    client.balanse = parseFloat(this.balanseNewRef.value);

    if (client.balanse <= 0) {
      client.status = 'blocked';
    } else {
      client.status = 'active';
    }

    mobileEvents.emit('savingClient', client);
    //переключаем отображение на показ 
    this.setState({ workMode: MobileClient.WORKMODE_SHOW });
  }

  handlerSaveClient = () => {
    //копия клиента из state
    let client = { ...this.state.clientInfo };


    client.lastName = this.lastNameRef.value;
    client.firstName = this.firstNameRef.value;
    client.secondName = this.secondNameRef.value;
    client.balanse = parseFloat(this.balanseRef.value);

    if (client.balanse <= 0) {
      client.status = 'blocked';
    } else {
      client.status = 'active';
    }

    mobileEvents.emit('savingClient', client);
    //переключаем отображение на показ 
    this.setState({ workMode: MobileClient.WORKMODE_SHOW });

  }

  handlerEditClient = () => {
    this.setState({ workMode: MobileClient.WORKMODE_EDIT });
  }
  handlerEditCancel = () => {
    this.setState({ workMode: MobileClient.WORKMODE_SHOW });
  }

  setTextInput = (element) => {
    this.newTextRef = ref;
  }


  render() {
    console.log('MobileClient render. ID - ' + this.state.clientInfo.id);


    return (
      <Fragment>
        {

          (this.state.workMode == MobileClient.WORKMODE_SHOW) &&
          <tr className='MobileClient'>
            <td >{this.state.clientInfo.lastName}</td>
            <td >{this.state.clientInfo.firstName}</td>
            <td >{this.state.clientInfo.secondName}</td>
            <td >{this.state.clientInfo.balanse}</td>
            <td className={classNames({ active: this.state.clientInfo.balanse > 0, blocked: this.state.clientInfo.balanse <= 0 })} >{this.state.clientInfo.status}</td>
            <td><Button onClick={this.handlerEditClient}>Редактировать</Button></td>
            <td><Button onClick={()=>{mobileEvents.emit('deletingClient', this.state.clientInfo.id)}}>Удалить</Button></td>
          </tr>

        }
        {
          (this.state.workMode == MobileClient.WORKMODE_EDIT) &&
          <tr className='MobileClient'>
            <td><input type="text" ref={this.setLastNameRef} defaultValue={this.state.clientInfo.lastName} /></td>
            <td><input type="text" ref={this.setFirstNameRef} defaultValue={this.state.clientInfo.firstName} /></td>

            <td><input type="text" ref={this.setSecondNameRef} defaultValue={this.state.clientInfo.secondName} /></td>
            <td><input type="number" ref={this.setBalanseRef} defaultValue={parseFloat(this.state.clientInfo.balanse)} /></td>
            <td> </td>
            <td><Button onClick={this.handlerEditCancel}>Отмена</Button></td>
            <td><Button onClick={this.handlerSaveClient}>Сохранить</Button></td>
          </tr>
        }
        {

          (this.state.workMode == MobileClient.WORKMODE_NEW) &&
          <div className="MobileClient-new">
            <label className="MobileClient-new__input">Фамилия
              <input type="text" ref={this.setLastNameNewRef} defaultValue="" />
            </label>
            <label className="MobileClient-new__input">Имя
              <input type="text" ref={this.setFirstNameNewRef}  defaultValue="" />
            </label>
            <label className="MobileClient-new__input">Отчество
              <input type="text" ref={this.setSecondNameNewRef}  defaultValue="" />
            </label>
            <label  className="MobileClient-new__input">Баланс
              <input type="number" ref={this.setBalanseNewRef}  defaultValue="" />
            </label>
            <br />
            <Button onClick={this.handlerSaveNewClient}>Сохранить</Button>
            <Button onClick={()=>{mobileEvents.emit('addNewClientCancel')}}>Отмена</Button>
          </div>
        }
      </Fragment>

    )
  }
}

export default MobileClient;