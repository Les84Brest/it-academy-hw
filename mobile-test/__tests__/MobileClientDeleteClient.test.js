"use strict";

import React from 'react';
import renderer from 'react-test-renderer';

import MobileClient from '../components/MobileClient';
import Button from "../components/Button";

test('MobileClient удаление клиента', () => {
  // исходные данные для рендера компонента
  const data = {

      "company": "MTS",
      "lastName": "Безустанный",
      "firstName": "Константин",
      "secondName": "Константинович",
      "balanse": -8.5,
      "id": 125,
      "status": "blocked"
    
  }
  // создаем экземпляр компонента для теста
  const component = renderer.create(
    <MobileClient clientInfo={data} />
  );

  let componentTree =  component.toJSON();
  expect(componentTree).toMatchSnapshot();
  
  let btnDeleteClient = component.root.find(el => el.props.children == "Удалить");
  
  expect(btnDeleteClient.props.onClick()).toBe(125);


});