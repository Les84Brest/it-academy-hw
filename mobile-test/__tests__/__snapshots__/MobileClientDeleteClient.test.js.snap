// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileClient удаление клиента 1`] = `
<tr
  className="MobileClient"
>
  <td>
    Безустанный
  </td>
  <td>
    Константин
  </td>
  <td>
    Константинович
  </td>
  <td>
    -8.5
  </td>
  <td
    className="blocked"
  >
    blocked
  </td>
  <td>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      Редактировать
    </button>
  </td>
  <td>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      Удалить
    </button>
  </td>
</tr>
`;

exports[`MobileClient удаление клиента 2`] = `
<tr
  className="MobileClient"
>
  <td>
    Безустанный
  </td>
  <td>
    Константин
  </td>
  <td>
    Константинович
  </td>
  <td>
    -8.5
  </td>
  <td
    className="blocked"
  >
    blocked
  </td>
  <td>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      Редактировать
    </button>
  </td>
  <td>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
    >
      Удалить
    </button>
  </td>
</tr>
`;
